docs/source/developer/iterating_expression.rst

```diff
@@ -0,0 +1,234 @@
+    {
+        using inner_shape_type = promote_shape_t<typename std::decay_t<CT>::shape_type...>;
+        using const_stepper = xfunction_stepper<F, R, CT...>;
+        using stepper = const_stepper;
+    };
+
+Then the inheriting class must define the following methods:
+
+.. code::
+
+    template <class S>
+    const_stepper stepper_begin(const S& shape) const noexcept;
+    template <class S>
+    const_stepper stepper_end(const S& shape, layout_type l) const noexcept;
+
+If the expression class inherits from ``xiterable`` instead of ``xconst_iterable``, the non-const counterparts
+of the previous methods must also be defined. Every method implemented in one of the base class eventually calls
+one of these stepper methods, whose mechanics is explained hereafter.
+
+Steppers
+~~~~~~~~
+
+Steppers are the low-level tools for iterating over expressions. They provide a raw API for "stepping" of a given
+amount in a given dimension, dereferencing the stepper, and moving it to the beginning or the end of the expression:
+
+.. code::
+
+    reference operator*() const;
+    
+    void step(size_type dim, size_type n = 1);
+    void step_back(size_type dim, size_type n = 1);
+    void reset(size_type dim);
+    void reset_back(size_type dim);
+
+    void to_begin();
+    void to_end(layout_type l);
+
+The ``reset`` and ``reset_back`` methods are shortcut to ``step_back`` and ``step`` called with ``dim`` and ``shape[dim] - 1``.
+The steppers are initialized with a "position" (that may be an index, a pointer to the underlying buffer of an container-based
+expression, etc...) in the expression, and can then be used to browse the expression in any direction:
+
+.. image:: stepper_basic.svg
+
+In this diagram, the data are stored in row-major order, and we step in the second dimension (dimension index starts at 0).
+The positions of the stepper are represented by the red dots.
+
+The ``to_end`` method takes a layout parameter, because the ending positions of a stepper depend on the layout used to iterate.
+Indeed, if we call ``step_back`` after a call to ``to_end``, we want the stepper to point to the last element. To ensure this
+for both row-major and column-major iterations, the ending positions must be set as shown below:
+
+.. image:: stepper_to_end.svg
+
+The red dots are the position of a stepper iterating in column-major while the green ones are the positions of a stepper iterating
+in row-major. Thus, if we assume that ``p`` is a pointer to the last element (the square containing 11), the ending positions
+of the stepper are ``p + 1`` in row-major, and ``p + 3`` in column-major.
+
+A stepper is specific to an expression type, therefore implementing a new kind of expression usually requires to implement a new
+kind of stepper. However `xtensor` provides a generic ``xindexed_stepper`` class, that can be used with any kind of expressions.
+Though not optimal, this solution allows to defer the implementation of a specific stepper.
```

Comment:
Even though it is generally not optimal, authors of new expression types can make use of the generic index stepper in a first implementation.